filebeat/input/filestream/internal/input-logfile/harvester_test.go

```diff
@@ -0,0 +1,331 @@
+		wg.Wait()
+		// wait until goroutine that started `harvester.Run` is finished
+		gorountineChecker.waitOriginalCount()
+
+		require.Equal(t, 1, mockHarvester.runCount)
+
+		// when finished removed from bookeeper
+		_, ok := hg.readers.table[source.Name()]
+		require.False(t, ok)
+		// stopped source can be stopped
+		require.Nil(t, hg.StopGroup())
+
+		gorountineChecker.waitOriginalCount()
+	})
+
+	t.Run("assert a harvester can be stopped and removed from bookkeeper", func(t *testing.T) {
+		mockHarvester := &mockHarvester{onRun: blockUntilCancelOnRun}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+
+		gorountineChecker := newGoroutineChecker()
+
+		hg.Start(input.Context{Logger: logp.L(), Cancelation: context.Background()}, source)
+
+		// run commands while harvester is running
+		gorountineChecker.waitWhileMoreGoroutinesWithFunc(func() {
+			// wait until harvester is started
+			if mockHarvester.runCount == 1 {
+				// assert that it is part of the bookkeeper before it is stopped
+				_, ok := hg.readers.table[source.Name()]
+				require.True(t, ok)
+				// after started, stop it
+				hg.Stop(source)
+				// wait until the harvester is stopped
+				gorountineChecker.waitOriginalCount()
+			}
+		})
+
+		// when finished removed from bookeeper
+		_, ok := hg.readers.table[source.Name()]
+		require.False(t, ok)
+	})
+
+	t.Run("assert a harvester for same source cannot be started", func(t *testing.T) {
+		mockHarvester := &mockHarvester{onRun: blockUntilCancelOnRun}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		inputCtx := input.Context{Logger: logp.L(), Cancelation: context.Background()}
+		gorountineChecker := newGoroutineChecker()
+
+		hg.Start(inputCtx, source)
+		hg.Start(inputCtx, source)
+
+		gorountineChecker.waitWhileMoreGoroutinesWithFunc(func() {
+			// error is expected as a harvester group was expected to start twice for the same source
+			err := hg.StopGroup()
+			require.Error(t, err)
+		})
+
+		require.Equal(t, 1, mockHarvester.runCount)
+	})
+
+	t.Run("assert a harvester panic is handled", func(t *testing.T) {
+		mockHarvester := &mockHarvester{onRun: panicOnRun}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		defer func() {
+			if v := recover(); v != nil {
+				t.Errorf("did not recover from harvester panic in defaultHarvesterGroup")
+			}
+		}()
+
+		gorountineChecker := newGoroutineChecker()
+
+		hg.Start(input.Context{Logger: logp.L(), Cancelation: context.Background()}, source)
+
+		require.Nil(t, hg.StopGroup())
+		gorountineChecker.waitOriginalCount()
+	})
+
+	t.Run("assert a harvester error is handled", func(t *testing.T) {
+		mockHarvester := &mockHarvester{onRun: errorOnRun}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		gorountineChecker := newGoroutineChecker()
+
+		hg.Start(input.Context{Logger: logp.L(), Cancelation: context.Background()}, source)
+
+		gorountineChecker.waitOriginalCount()
+
+		_, ok := hg.readers.table[source.Name()]
+		require.False(t, ok)
```

Comment:
reoccuring pattern consider:

```
func requireHasSource(t *testing.T, g ..., source Source) {
}

func requireHasNotSource(t *testing.T, g ..., source Source) {
}
```

The test checks some implementation detail. Having the assert encapsulated in a single function will require us to change only a single place (without touching the actual test, hopefully) in case some internals would change.